ls1intum/Artemis
# Tutors and meta info for teams

### Checklist
- [x] Client: I added multiple integration tests (Jest) related to the features (updated existing)

### Motivation and Context
The way the team-based exercise workflow is intended to work for the EIST course next semester, it is important that teams belong to tutors (who will create the teams and assess their submissions). This PR enables the following behavior: When a tutor creates a team, he will be the owner of the team. Only instructors can re-assign tutors to teams after that. All TAs can see which team belongs to which tutor. Students can see their team tutor on the team detail page and can also contact him/her via email. Additionally, the team detail page features some meta information on when the team was created and when it was last modified.

### Description
- Make the entity `Team` extend `AbstractAuditingEntity` (adds the fields `createdBy`, `createdDate`, `lastModifiedBy`, `lastModifiedDate` which are populated automatically)
- Add the field `owner` to the entity `Team` (which is a foreign key to the users table, this will usually be a tutor)
- Make sure that only instructors can modify the owner of a team (both server-side and client-side)
- Add the field "Tutor" to the team edit dialog (all tutors for course are loaded from server and can be selected; the field is read-only for TAs)
- Add column "Tutor" to teams overview for exercise (shows tutor name and has `mailto` link except for admins it links to the user management page)
- Add subline with meta information under the headline on the team detail page (contains the info: tutor of team, created date and last modified date; admins can also see the principal of the users who created and last modified the team)

### Steps for Testing
1. Log in to Artemis
2. Navigate to Course Management
3. Click on Exercises for a Course
4. Click on Teams for a team-based exercise (or create one first)
5. Create a few teams
6. Check that new "Tutor" column exists (might be empty for existing teams)
7. Click on a team short name to go to the team detail page
8. Check that new meta info is available below headline (might not be correct for teams that existed before this PR)
9. As instructor, click on "Edit" to open the team edit dialog
10. Change the team tutor and verify that the update worked
11. Verify that changing the team tutor is not possible as a TA

### Screenshots
1. Team detail page (as `TA / INSTRUCTOR`): New row of information below headline (tutor + created + optional last modified) => tutor name has a `mailto` link so he can be contacted via email
![OnPaste 20200331-142007](https://user-images.githubusercontent.com/7109225/78025898-1634d200-735b-11ea-8ed6-2756e7f4f89a.png)

2. Team detail page (as `TA / INSTRUCTOR`): Created and Last Modified timestamps appear as a tooltip on hover to see the exact time
![OnPaste 20200331-142033](https://user-images.githubusercontent.com/7109225/78025903-16cd6880-735b-11ea-8ab0-2bd79a3e7e55.png)

3. Team detail page (as `ADMIN`): Tutor name is now linked to `/user-management` detail page + created by and last modified by are exposed and also linked to user page
![OnPaste 20200331-142113](https://user-images.githubusercontent.com/7109225/78025905-1765ff00-735b-11ea-964d-f1790f2e7019.png)

4. Team edit dialog (as `TA`): New field "Tutor" that is read-only for tutors. A tooltip informs the tutor that a re-assignment can only be performed by an instructor. This way a tutor knows that a re-assignment is possible by contacting the instructor and requesting it. (The instructor sees the same tooltip so that he is aware that tutors can't do this on their own and thus might request it.)
![OnPaste 20200331-142140](https://user-images.githubusercontent.com/7109225/78025908-17fe9580-735b-11ea-9db0-c804027d657d.png)

5. Team edit dialog (as `INSTRUCTOR`): New field "Tutor" can be edited to re-assign the team to a different tutor. Initially, the tutor who created the team is its owner. When the instructor clicks into the Tutor field, all tutors for that course are loaded from the server (only the first time he clicks on it, reused afterwards for the duration of the dialog being open), so there is no need to guess who is a tutor by performing searches (and since the number of tutors per course isn't very high, this is acceptable from a performance standpoint). Autocomplete is still supported though obviously.
![OnPaste 20200331-142208](https://user-images.githubusercontent.com/7109225/78025910-18972c00-735b-11ea-972b-9aa6e71f8851.png)

6. Teams overview for exercise (as `INSTRUCTOR`): New column "Tutor" that shows who the tutor is for each team. TAs can also see this. The search bar above will also find teams when someone searches for the tutor name or login.
![OnPaste 20200331-142235](https://user-images.githubusercontent.com/7109225/78025912-18972c00-735b-11ea-9088-cf47cffae17a.png)
